src/dictionary.c

```diff
@@ -160,6 +158,24 @@ void dict_put(struct dictionary* d, uint32_t hash, R_len_t i) {
   d->used++;
 }
 
+// Assume worst case, that every value is distinct, aiming for a load factor
+// of at most 77%. We round up to power of 2 to ensure quadratic probing
+// strategy works. Maximum power of 2 we can store in a uint32_t is 2^31,
+// as 2^32 is 1 greater than the max uint32_t value, so we clamp sizes that
+// would result in 2^32 to INT_MAX to ensure that our maximum ceiling value
+// is only 2^31. This will increase the load factor above 77% for `x` with
+// length greater than 1653562409 (2147483648 * .77), but it ensures that
+// it can run.
+static inline
+uint32_t dict_key_size(SEXP x) {
+  uint32_t size = (uint32_t)(vec_size(x) / 0.77);
```

Comment:
While we are at it we should probably make that code robust to a switch to long vector sizes:

- Check that the division doesn't produce `NaN` by overflow.
- Check that the division result (a double since literal floats are always doubles and integers are promoted to the float type) is representable as a `u32`, i.e. it is equal or less to `2^53`.

Reply:
For long vectors, I guess the largest size would be the max `ptrdiff_t` value, usually `9223372036854775807`? If you divide that by `.77`, that won't produce a `NaN`. Did you have something else in mind?

> representable as a u32, i.e. it is equal or less to 2^53.

Did you mean `2^32 - 1`? The max u32 value? If we were using u64 hash values, then yea I agree the max double value where all integerish values smaller than it are precisely representable as an integer is `2^53`, but I'm not sure we want that here.

What do we do in the case where we get a long vector that this implementation can't handle? Error?

Reply:
(1) oops you're right long vectors can't be that long. (2) You're also right, I was confused and thinking about u64 while writing u32.

An error seems appropriate. If that becomes a problem we can add a 64 bits hashing code path later on.